core/domain/learner_progress_services.py

```diff
@@ -0,0 +1,415 @@
+            exploration.
+        exploration_version: str. The version of the exploration played by the
+            learner.
+    """
+    incomplete_activities_model = (
+        user_models.IncompleteActivitiesModel.get(
+            user_id, strict=False))
+    if not incomplete_activities_model:
+        incomplete_activities_model = (
+            user_models.IncompleteActivitiesModel(id=user_id))
+
+    completed_exploration_ids = get_all_completed_exp_ids(user_id)
+
+    subscribed_exploration_ids = (
+        subscription_services.get_exploration_ids_subscribed_to(user_id))
+
+    incomplete_activities = get_incomplete_activities_from_model(
+        incomplete_activities_model)
+
+    if (exploration_id not in completed_exploration_ids and
+            exploration_id not in subscribed_exploration_ids):
+
+        if exploration_id not in incomplete_activities.incomplete_exploration_ids: # pylint: disable=line-too-long
+            incomplete_activities.add_exploration_id(exploration_id)
+
+        last_playthrough_information_model = (
+            user_models.ExpUserLastPlaythroughModel.get(
+                user_id, exploration_id))
+        if not last_playthrough_information_model:
+            last_playthrough_information_model = (
+                user_models.ExpUserLastPlaythroughModel.create(
+                    user_id, exploration_id))
+
+        last_playthrough_information = get_last_playthrough_information(
+            last_playthrough_information_model)
+        last_playthrough_information.update_last_played_information(
+            exploration_version, state_name)
+
+        save_last_playthrough_information(last_playthrough_information)
+        save_incomplete_activities(incomplete_activities)
+
+
+def mark_collection_as_incomplete(user_id, collection_id):
+    """Adds the collection id to the list of collections partially completed by
+    the user unless the collection has already been completed or has been
+    created/edited by the user or is already present in the incomplete list.
+
+    Args:
+        user_id: str. The id of the user who partially completed the collection.
+        collection_id: str. The id of the partially completed collection.
+    """
+    incomplete_activities_model = (
+        user_models.IncompleteActivitiesModel.get(user_id, strict=False))
+    if not incomplete_activities_model:
+        incomplete_activities_model = (
+            user_models.IncompleteActivitiesModel(id=user_id))
+
+    completed_collection_ids = get_all_completed_collection_ids(user_id)
+
+    subscribed_collection_ids = (
+        subscription_services.get_collection_ids_subscribed_to(user_id))
+
+    incomplete_activities = get_incomplete_activities_from_model(
+        incomplete_activities_model)
+
+    if (collection_id not in subscribed_collection_ids and
+            collection_id not in incomplete_activities.incomplete_collection_ids and # pylint: disable=line-too-long
```

Comment:
You can break after "in".

That said, I'm wondering whether our property names are too long. I'll comment below.